### ISSUE TITLE
Separate version out into a requested hint?

### ISSUE THREAD

**Issue by mikewest** [WICG/ua-client-hints#7]
What if we dropped the granularity of the defaults even more by splitting the version out into a distinct hint, and leaving UA as just the brand? That is, we'd send the following by default:

```
Sec-UA: "Chrome", "Not Chrome"
Sec-UA-Mobile: ?0 // Maybe, depending on #5
```

And the following after opt-in via `Accept-CH: UA-Version, UA-Platform`:

```
Sec-UA: "Chrome", "Not Chrome"
Sec-UA-Mobile: ?0
Sec-UA-Version: 77.0.5000.123
Sec-UA-Platform: "Linux 18000.23.2311.3"
```

@amtunlimited @yoavweiss

**Comment by amtunlimited** [comment]
I wouldn't be opposed. We could also put out "Canary | Dev | Beta | Stable | Other(?)" to give some sort of compat information, but I don't know how useful that could be. I also don't know how useful, in general, the version number is in the every day life of a developer.

**Comment by yoavweiss** [comment]
Do we have a summary of use cases for which we are sending *any* UA information without opt-in?

I thought that the current brand+version was aiming at enabling developers to bypass version specific compat bugs. Removing the version would mean that the use-case is not enabled, which makes me wonder what is.

**Comment by mikewest** [comment]
With the _very_ large caveat that I have no data on this, and that it's all poorly-informed intuition and speculation:

The thing that seems _critical_ to me, that I'd expect the overwhelming majority of developers to opt-into, is the brand with which a particular user agent claims itself to be compatible. That gives an indication of the rendering engine, and a baseline understanding of support (or lack thereof!) for a broad set of capabilities. That seems core to the decision of what content to send in a way that every other piece of information doesn't.

The broad "mobile" or "not" distinction also seems widely applicable.

I do think there are clear cases in which other pieces of information are helpful. Bugs that I've personally written have been worked-around via UA string checks on the browser's (major) version. I'm sure specific use cases could be found for the rest of the bits of the UA string as well. We should offer them to developers! As opt-ins!

**Comment by yoavweiss** [comment]
The "mobile or not" use case is clear to me, but seems separate from "brand", right?

Otherwise, can you expand on the use-case you have in mind? What kind of content differences do we expect developers to send to "Chrome" vs. "Firefox" vs. "Safari" that feature detection cannot cover in better ways?

**Comment by mikewest** [comment]
> The "mobile or not" use case is clear to me, but seems separate from "brand", right?

> Otherwise, can you expand on the use-case you have in mind? What kind of content differences do we expect developers to send to "Chrome" vs. "Firefox" vs. "Safari" that feature detection cannot cover in better ways?

Client-side feature detection is great. We should encourage it. Client-side feature detection has performance costs, or so I'm told over and over again by server-side folks, and the friendly chairs of the Web Perf working group. :) We should ask actual web developers at some point, but I expect the rendering engine to be an important consideration when deciding what HTML/JavaScript/CSS to send.

(Also, I expect that "Chrome" vs "Firefox" vs "Safari" is detectable purely by examining the network stacks' characteristics, especially after a TLS handshake, so I don't think it creates any _actual_ addition of entropy.)

**Comment by yoavweiss** [comment]
> Client-side feature detection has performance costs, or so I'm told over and over again by server-side folks

Might be interesting to talk to them and better understand their use cases.

My concern is that exposing "brand" by default can increase UA blocking (which is a pattern I believe we want to discourage), or other server-side assumptions (e.g. "Firefox doesn't support feature X *right now*, so let's not send that code to it, and ignore the fact that they might in the future").

Looking into the use-cases can also help us to identify ways in which feature detection can be made faster (e.g. maybe conditional JS loading can help the "only load this JS if feature X is supported" case).

**Comment by amtunlimited** [comment]
I'd also like to point out that "sec-ch-ua" is the only client hint (at least in this set) where the "format" changes, at least enough that any one format string or reg-ex doesn't fix without feeling a little kluge-y. I wouldn't mind splitting out the full version number into a separate client hint in either case. Should that be it's own issue?

**Comment by domenic** [comment]
I'd strongly suggest splitting the header in two for the simple reason that it avoids potentially-buggy parsing code.

Consider the current single-giant-string world. Popular libraries use regexes, etc. to parse it into components they care about, such as version, engine, brand, mobile/not, etc. This parsing is sometimes buggy, leading to famous issues such as Opera 10 being unable to send an accurate version number because of code that assumes anything starting with `Opera 1` is version 1 of Opera.

This proposal almost completely obviates the need for such potentially-buggy parsing code. But it doesn't go quite far enough! In particular it still requires parsing code on the `Sec-CH-UA` header to separate out the brand and the version. The first thing that popped into my head is that this architecture is very likely to prevent multi-word user agent brands (such as the spec's examples of `The World’s Best Web Browser` or `Examplary Browser`) from ever becoming web compatible, because we'll see a lot of parsing code that looks like

```js
const [brand, version] = secUACHHeader.split(' ');
```

To summarize, it seems like a missed opportunity to remove a whole class of bugs, if you just went one step further and separated out all distinct pieces of information into their own header.

**Comment by sgomes** [comment]
As this suggestion is considered, I'd like to emphasise the need for a coarse browser version (major version should be sufficient) in the first request, due to latency-sensitive contexts such as differential build serving.

Since there doesn't exist a generic* client side approach for differential build serving that doesn't require an extra roundtrip, it's useful to rely on the information in UA strings to determine whether it's possible to serve a smaller build vs a more comprehensively polyfilled, albeit larger build.

Moving the major version from one first request header to another would be fine, but pushing it to the second request would invalidate any attempt in avoiding the extra roundtrip.


* – module/nomodule isn't particularly generic, as it's attached to an increasingly irrelevant arbitrary point in time

**Comment by yoavweiss** [comment]
I agree with both @domenic and @sgomes - We should split out the version into its own separate field (for ergonomics reasons that @domenic pointed out), but continue to send it by default, as the use case of bug workarounds or differential serving requires the browser's meaningful version.

**Comment by amtunlimited** [comment]
I see a few options:

1. Keep `sec-ch-ua` the same, separate the full version to a deferred hint:
```
Sec-CH-UA: "Browser 100"
...
Accept-CH: version
...
Sec-CH-UA: "Browser 100"
Sec-CH-Version: "100.1.2"
```

2. Remove the version from `sec-ch-ua`, make it a different hint that you can have a accept-ch for it:
```
Sec-CH-UA: "Browser"
Sec-CH-Version: "100"
...
Accept-CH: version
...
Sec-CH-UA: "Browser"
Sec-CH-Version: "100.1.2"
```

Ergo-wise I kinda prefer option 1 as it doesn't change the "format" (read: regex) of any headers between requests. When we get to a point where we start denying access to hints, which the server gets no notification of, it doesn't know which regex to use, making an even grosser regex or leading to bugs.

**Comment by yoavweiss** [comment]
Another option is:
```
Sec-CH-UA: "Browser"
Sec-CH-Version: "100"
...
Accept-CH: full-version
...
Sec-CH-UA: "Browser"
Sec-CH-Version: "100"
Sec-CH-Full-Version: "100.1.2"
```

**Comment by mikewest** [comment]
The multi-header model becomes more difficult when considering that `Sec-CH-UA` is a list. Relying on mapping position of brand and version between two lists is somewhat ugly.

We could instead represent version as a parameter on the brand? That is, `Sec-CH-UA: "Browser"; version=100, "Another"; version=1000, "etc"; version=0`.

**Comment by yoavweiss** [comment]
> We could instead represent version as a parameter on the brand? That is, `Sec-CH-UA: "Browser"; version=100, "Another"; version=1000, "etc"; version=0`

@mikewest - That's indeed where I ended up! (only I reached that conclusion independently because I'm bad at email...)
Main difference is that I also made `version` a string, as I didn't want to limit future browsers to necessarily use integers as their versions. (I also named it `v` because compression)

> I'd strongly suggest splitting the header in two for the simple reason that it avoids potentially-buggy parsing code.

@domenic - I landed on the following:
* The HTTP header is a sh-list, where each sh-item is a string with a "v" parameter that indicates the version. That enables servers to reuse standard header parsing code (using [Structured Headers](https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html)).
* On the JS side of things, the UA is an array of dictionaries, each with a `brand` and `version`. So no parsing required.

Let me know what you think.

Going back to the OP, I think we have a lot of feedback on the list (as well as from @sgomes here) that the significant version number is important in order to only serve polyfills to browsers that need them. So we should keep it exposed by default, at least until we'd have a solution for the "getting hints on the very-first-view problem".

Closing this, as I believe it was addressed, but feel free to reopen if you feel something here requires further discussion.
